Keep manager (#802)
enarx/enarx: host-components/README.md

# How to run the host-components

MikeCamel@github.com

2020-08-11

This file explains the steps you need to run the Enarx host-components
as a Proof of Concept demo.  The exact capabilities of this demo will
change over time, and so this document should be updated to reflect that.

## What the demo shows

The demo is made up of several components:
- a keep-manager
- multiple keep-loaders, created by the by the keep-manager using
systemctl
- an app-loader, which waits for, and then runs, a WebAssembly file
- a keep-manager-tester, which talks to the keep-manager, requesting
information, setting up the keep-loaders and preparing them to accept
the WebAssembly file
- an app-loader-tester, which sends the app-loader a WebAssembly file

## Preparing the demo

>  - The demo currently only runs on a single machine, as it listens on
> localhost.  Other hosts are supported, but HTTPS certificates do not work.

Reviewer: ```suggestion
-  The demo has currently only been tested on Fedora 32.
 - The demo currently only runs on a single machine, as it listens on
localhost.  Other hosts are supported, but HTTPS certificates do not work.
```

Author: Done.

To change the host, edit `LOCAL_LISTEN_ADDRESS` in
`enarx/host-components/src/lib.rs`.  To test, changes are required in the files
`enarx/host-components/src/keep-manager-tester.rs` and
`enarx/host-components/src/app-loader-tester.rs`
 - The demo currently has no support for Keeps running in TEEs: it supports
wasmtime.
 - The demo currently has little support for outputs from a workload.
 - The demo currently doesn't support filesystem access from a workload.
 - The demo currently doesn't support socket access from a workload.
 - The demo currently has a hard-coded .wasm (WebAssembly) file to run.
 Others can be used, but do not expect much output!

JSON is used for communications.
Communications to the keep-manager from the keep-manager-tester are over HTTPS.
Communications to the app-loader from the app-loader-tester are over HTTPS
with both client and server certificates (the latter being created and self-
signed by the app-loader - see note above about non-localhost hosts).

In order to run the demo, you shouldn't need any tools beyond those normally
required to compile Enarx.

Steps:
1. download the latest source code from github (the easiest way is to do a

> `git clone` of the repository.

Reviewer: Until this is merged, the easiest way to get this branch locally is to clone locally with:
`git clone --single-branch --branch keep-manager https://github.com/MikeCamel/enarx/`

2. enter the enarx/keep-runtime directory
3. run `cargo build`
4. enter the enarx/host-components directory

> 5. edit the file `lib.rs` to reflect the location on your system of the
> `keep-runtime` binary created in previous

Reviewer: ```suggestion
5. edit the file `enarx/host-components/src/lib.rs` to reflect the location 
on your system of the `keep-runtime` binary created in the previous step
(it should been placed at enarx/keep-runtime/target/x86_64-unknown-linux-musl/debug/keep-runtime)
```

Author: Fixed.

6. run `cargo build`

> 7. enter the enarx/host-components directory

Reviewer: This is the directory we entered at step 4 and haven't left since, and given we won't be doing any further `cargo build`ing, we don't really need to enter any directory: we can just edit the files directly from where we are.

This has the added benefit of making the instructions lighter to remove some steps.

```suggestion
```

Author: Made some changes to allow this to work, then removed step 7 & renumbered.

> 8. edit the file `enarx-keep@.service` to update the `ExecStart` entry

Reviewer: ```suggestion
7. edit the file `enarx/host-components/external/enarx-keep@.service` to update a few entries
7.1 the `ExecStart` entry, to reflect the location of the `keep-loader` binary created in the previous
step (it should been placed at enarx/host-components/target/debug/keep-manager)
```

Reviewer: Another thing here, it would be useful to explain why a systemd unit is used here: what it does, how necessary it is etc.
Unavoidably, someone will ask if they can use sysv instead and explaining more is "always" useful.

Author: Done: added a number of notes here and improved relative paths.

> to reflect the location of the `keep-loader` binary created in the previous
> step

Reviewer: ```suggestion
```

Author: not sure what the change requested is.

> 9. edit the file `enarx-keep@.service` to update the `StandardOutput` and
> `StandardError` entries to reflect your preferred locations for the files to
> record stdio and stderr

Reviewer: ```suggestion
7.2. the `StandardOutput` and`StandardError` entries to reflect your preferred locations for the files to
record (log) stdio and stderr
```

Author: think it's confusing to have multiple list levels, so I'm leaving it.

> 10. run `sudo ln -s enarx-keep@.service /etc/systemd/user/enarx-keep@.service`

Reviewer: ```suggestion
8. run `sudo ln -s enarx-keep@.service /etc/systemd/user/enarx-keep@.service`
```

Reviewer 2: > Can't install the systemd unit, as i don't have sudo there.

As the whole stuff is not (yet) installable state, it might be more courteous to suggest installing the unit file under user's home directory. `man systemd.unit` mentions a couple of locations: `~/.config/systemd/user/*` and `~/.local/share/systemd/user/*`.

Author: I tried a number of options here, and struggled to get them to work, despite the man pages' assurances.  I suggest we leave for now.

Reviewer 2: Hmm, from my testing, the following steps (without sudo) work for me:
```console
$ mkdir -p ~/.config/systemd/user
$ cp external/enarx-keep@.service ~/.config/systemd/user
$ systemctl --user enable enarx-keep@
$ systemctl daemon-reload --user
```

> 11. run `sudo systemd reload-daemon` (running this with the parameter `--user`

Reviewer: ```suggestion
9. run `sudo systemd reload-daemon` (running this with the parameter `--user`
```

Author: renumbering done.

Reviewer 2: The command here should be `systemctl daemon-reload`.

Steps:
1. download the latest source code from github (the easiest way is to do a
`git clone` of the repository.
2. enter the enarx/keep-runtime directory
3. run `cargo build`
4. edit the file `enarx/host-components/src/lib.rs` to reflect the location
on your system of the `keep-runtime` binary created in the previous
step (it should be have been created in
`$YOUR_ROOT/enarx/keep-runtime/target/x86_64-unknown-linux-musl/debug/keep-runtime`)
5. enter the enarx/host-components directory
6. run `cargo build`
7. edit the file `enarx/host-components/external/enarx-keep@.service` to
update the `ExecStart` entry to reflect the location of the `keep-loader`
binary created in the previous step
8. edit the file `enarx/host-components/external/enarx-keep@.service` to
update the `StandardOutput` and `StandardError` entries to reflect your
preferred locations for the files to record stdio and stderr
9. run `sudo ln -s enarx-keep@.service /etc/systemd/user/enarx-keep@.service`
10. run `sudo systemd reload-daemon` (running this with the parameter `--user`
**may** work)

If you have run the demo before, you should kill old instances
of the keep-loader.  This is best down by running the command

> `pkill -9 keep-loader`.  You may also wish to delete any keeploader output

Reviewer 3: Instead of pkill, you might want to mention the proper systemd way.

```console
$ systemctl stop 'enarx-keep@*'
```

Author: Because it doesn't seem to work!

Author: Don't think this will get fixed in time for the end of the Sprint, so new issue here for all systemd-related questions: https://github.com/enarx/enarx/issues/849

Author: Fixed.

files (the locations of which you set in step 8).  Although it's unlikely to
have any impact on the demo, you may wish to clean up old files with
`rm /tmp/enarx-keep*.sock`.

> **NOTE** if you make any changes to the keep-loader.rs file, after recompiling

Reviewer 3: Huh? Why remove the link?

```console
$ systemctl try-restart 'enarx-keep@*'
```

should do the trick, if the service file didn't change.

Author: Don't think this will get fixed in time for the end of the Sprint, so new issue here for all systemd-related questions: https://github.com/enarx/enarx/issues/849

Author: Fixed.

it, you will need to remove the old link in /etc/systemd/user/, recreate it
(step 10 above), and perform reload the systemctl daemon (step 11 above).

## Running the demo

You will need two different command line terminals, both running on the same
machine.

Terminal 1 - server:
 - enter the `enarx/host-components` directory

>  - run the command `.target/debug/keep-manager`

Reviewer 3: I prefer:

```console
$ cargo run --bin keep-manager
```

which ensures, that you actually built the stuff :smile:

Author: Fair.  The reason I moved to the alternative was frankly because of the number of warnings I was getting each time I did that.  Those warnings are now fixed, so changing this!
